### PR TITLE
Container Registry VM Disks

### PR DESCRIPTION
**Pull request by davidvossel**
The ContainerRegistryDisk feature is being proposed as an alpha feature. The Corresponding proposal for this pull request can be found here #310. The issue associated with these patches can be found here #315

*NOTE: This feature shouldn't be seen as conflicting with the concepts proposed in #311. Once we have a long term strategy for how storage should be represented in KubeVirt, we may choose to revise the user facing parts of the ContainerRegistryDisk feature conform in a consistent way with respects to how we want KubeVirt represent storage in the VM definition. We are making no guarantees about backwards compatibility by introducing this alpha feature.*

# Overview
The **ContainerRegistryDisk:v1alpha** feature allows users to place VM disks in container images which can be stored in the container registry. From there, users can attach any number of disks stored in the container registry as local ephemeral storage for a VM.

For people familiar with AWS services, the pattern used here is very similar to AWS's [instance store-backed AMIs](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/InstanceStorage.html) 

With this pattern VM disks are stored somewhere reliably on the network *(In AWS's case this is S3, for KubeVirt this is the container registry)* and pulled in as local ephemeral storage for VMs to consume at launch.

While the local data is not persistent between VM restarts, there are some advantages to using this pattern over persistent network block storage.
* **Resiliency**: Active VMs continue to work despite an outage to the block storage service.
* **Simplicity**: Reduced operational complexity. No block storage required. There are hidden costs organizations pay for each component introduced into a distributed system. 

# User Example
**Place a bootable VM disk into the container registry**
```
cat << END > DockerFile
FROM kubevirt.io:registry-disk-v1alpha
ADD fedora25.qcow2 /disk
END

docker build -t vmdisks/fedora25:latest .
docker push vmdisks/fedora25:latest
```
**Attach VM disk as local ephemeral storage to a VM definition**
```
cat << END > vm.yaml
apiVersion: kubevirt.io/v1alpha1
kind: VM
metadata:
  name: testvm-ephemeral
spec:
  domain:
    devices:
      disks:
      - type: ContainerRegistryDisk:v1alpha
        source:
          name: vmdisks/fedora25:devel
        target:
          dev: vda
      interfaces:
      - source:
          network: default
        type: network
    memory:
      unit: MB
      value: 2048
    os:
      type:
        os: hvm
    type: qemu
END
```
**Launch the VM using the KubeVirt runtime**
```
kubectl create -f vm.yaml
```
# Testing
## Manual Testing
In our vagrant dev environment this feature can be tested by posting the cluster/vm-ephemeral.yaml file to a cluster.  This VM definition attaches the VM disk placed in the cirros-registry-disk-demo:devel container image, which is pre-built during 'make vagrant-deploy'
```
cluster/kubectl create -f cluster/vm-ephemeral.yaml
```
## Functional Testing
A new functional test is introduced to exercise the ContainerRegistryDisk feature. **/tests/registry_disk_tests.go**

# Implementation Details
* KubeVirt provides a base container that serves up VM disks as a iscsi network block device libvirt can consume. This base container is found in **/cmd/registry-disk-v1alpha**
* Users push their VM disks into the container registry using the base image. An example of this is found in **/images/cirros-registry-disk-demo**
* Users reference VM disks backed by the container registry with the **ContainerRegistryDisk:v1alpha** disk type when defining the VMs.
* **virt-controller** searches the VM definition for **ContainerRegistryDisk:v1alpha** disks and places containers for each disk in the virt-launcher pod. 
* The virt-launcher pod is launched with the the container backed disks and readiness checks are performed to determine when the disks are accessible.
* **virt-controller** places the VM on the node with the corresponding virt-launcher pod once all container disks' readiness checks pass.
* When **virt-handler** is defining the domain xml for a VM spec, the ContainerRegistryDisk:v1alpha disk type is converted to iscsi network disk (pointing to the disks served up by the containers) automatically.
* Anything the VM writes to the local ephemeral disks are destroyed when when the VM is torn down as a result of the virt-launcher pod being destroyed.

# Future Work
* I intentionally left out iscsi CHAP authentication from this initial pull request. If we decide to proceed with this feature, virt-controller should generate auth credentials automatically that ensure only the VM associated with the corresponding virt-controller pod can access the disks.
* Read-Only disks. 
* container pull policy (always, if present, ect...)
* More advanced functional tests need to be added that exercise failure conditions. 
* The initial feature set only includes support local ephemeral disks.  In the future we will consider the possibility of automating the dynamic provisioning of block storage for container registry disks for persistent storage. More thoughts on that can be found in this proposal. #311

### CODE REVIEW

File: pkg/virt-handler/vm.go [hunk 1]
```diff
@@ -291,6 +292,12 @@ func (d *VMHandlerDispatch) processVmUpdate(vm *v1.VM, shouldDeleteVm bool) erro
 		return err
 	}
 
+	// Map Container Registry Disks to block devices Libvirt can consume
+	vm, err = registrydisk.MapRegistryDisks(vm)
+	if err != nil {
+		return err
+	}
```
Comment (berrange) [inline pkg/virt-handler/vm.go]: This is an example of what I want us to avoid doing - it is hardcoding interaction with a specific disk technology. I would like to see this using the REST api approach I described, where it queryies the container at a designated end point to get the desired kubvirt.api.v1.DiskSource back in a generic manner.

Comment (davidvossel) [inline pkg/virt-handler/vm.go]: Right, so I don't exactly disagree or agree with your thoughts on the REST endpoint quite yet.  However, I think the REST endpoint is out of the scope of what I'm trying to do here though. 

This comes to the point I made at the top of this PR about this feature's lack of interference with your storage proposal #311.

I implemented this feature in a way that is most natural for KubeVirt as it is today. If you look about 3 or 4 lines up you'll see we map PVCs to specific technology backends. These are established patterns I'm using with clear entry points that are isolated to the ContainerRegistryDisk feature.

What I don't want to see is for this feature to be blocked by the much larger storage discussion.   

If we decide to move to a generic REST api for retrieving storage devices, then the ContainerRegistryDisk feature will migrate to use that logic.  I purposefully encapsulated the logic in this feature to make this type of transition simple for us.

Comment (berrange) [inline pkg/virt-handler/vm.go]: The other storage mechanisms that we integrate are different because they are the base layer "native" approaches to integration in k8s, where as this is adding a new mechanism which is  really a layer above the core storage backend infrastructure.

I'm don't seeing using REST as an item that blocks you - you could easily just implement a rest API in this patch series now. I would much rather see that generic approach taken straight away than to add this code that we know someone will have to rewrite later. Trying the REST approach straightaway, will also demonstrate the viability of that generic approach to dynamic storage integration

Comment (davidvossel) [inline pkg/virt-handler/vm.go]: > I'm don't seeing using REST as an item that blocks you - you could easily just implement a rest API in this patch series now. I would much rather see that generic approach taken straight away than to add this code that we know someone will have to rewrite later.

We are not certain that someone will have to re-write this later.  Likewise, we are not certain that introducing a REST api for this feature won't result in someone having to re-write the api part out later because it's found to be unnecessary.

The implementation, as it is now, is expressed naturally without the need for a REST api. If we see that REST api integration makes sense for this feature in the future, then we'll adapt.

> Trying the REST approach straightaway, will also demonstrate the viability of that generic approach to dynamic storage integration

I'd prefer not to attempt to prove the viability of a generic storage REST api *(that I'm not 100% sold on yet)* in a feature that doesn't need it.

I do however agree that the generic storage REST api is a good idea for us to continue discussing in a channel outside of this PR. I'd like for these to be treated as separate non conflicting discussions *(hence the disclaimer a the top of this PR about the #311 proposal)*

Comment (berrange) [inline pkg/virt-handler/vm.go]: My view is that we must define a generic storage framework first. I don't want us adding technical debt by integrating custom storage drivers into the core code in this way. This is why I put up the design proposal for a generic system to integrate storage. The generic system does not have to be very much extra work to achieve, so I see no compelling reason to do this docker specific solution first.

Comment (davidvossel) [inline pkg/virt-handler/vm.go]: > My view is that we must define a generic storage framework first. I don't want us adding technical debt by integrating custom storage drivers into the core code in this way. This is why I put up the design proposal for a generic system to integrate storage. The generic system does not have to be very much extra work to achieve, so I see no compelling reason to do this docker specific solution first.

I disagree that this is technical dept. From my perspective this is agile development. There's also no reason (from my perspective) that the generic abstraction has to come before this feature. We iterate on concepts as we gain a better understanding of the full picture.  I don't believe the REST api is a fully baked concept yet. The relationship between a generic storage api and this feature is not well understood (at least to me). When it is fully understood, we will adapt. It's possible this feature will even help demonstrate the usefulness of the api as we continue to have that discussion. 

Until then, I believe we benefit from this feature's functionality now.

There are efforts underway to prove out other parts of KubeVirt that require dynamically generated ephemeral disks just to properly POC and test. For example, nested k8s nodes, VMGroups, and cloud-init scripts. These all benefit from the ContainerRegistryDisk feature. This feature gives us the ability to begin iterating on these other concepts (in a tangible way) independently of the generic storage api.

Comment (berrange) [inline pkg/virt-handler/vm.go]: Calling it "agile" is just using a buzzword to justify creating technical debt by not considering the general storage picture. There's no time critical reason why this feature must be pushed in right now rather than providing a general solution, which is barely any more work than what's done here. so I'm still against merging this as it is proposed here.

Comment (rmohr) [inline pkg/virt-handler/vm.go]: @berrange I have problems to see, why directly doing something, instead of hiding it behind a REST API, where we don't know if we need it, what it would do, or how it would look like, would introduce a technical dept. Completely independent if one likes a specific buzzword or not.

I agree with @davidvossel, that something like this PR opens possibilities for easy to integrate epemeral VMs without having to care about storage. If this proposed feature turns out, to not make sense, we can easily remove it again, if it makes sense, but we re-architect the storage part significantly later on, all the functional tests are still valuable ...

Also one note in general, regarding "microservices". So far KubeVirt, tries to follow as much as possible the Kubernetes architecture, which is very simple: As it's core, one binary for controllers, one binary for api-server extensions, and one binary for the node daemon. Only if it makes sense to exchange something arbitrarily, by third party implementations, they provide hooks or registration mechanisms for extension.

Here also a short summary, what people wanted to have extra containers/services/APIs or external service depenencies for withing the last two weeks, as part of the core infrastructure in KubeVirt:

 * Extra API and service for networking
 * Extra API and service for storage
 * A dependency on KubeDNS for our core infrastructure
 * Providing console stream access through a proxy (not through virt-api like right now), like spice/vnc
 * Having an extra credential resolver container for spice/vnc/console between the proxy and virt-api (which we don't have right now)

My intent is not to say that it is wrong to have one or all of them. But let's just write proposals for another two weeks, and say yes to all of them, then we have between 20 or 30 services.

Long story short: From my view, if it is something which can be done on the node, virt-handler should do it. Having extra daemons on the node, is not something I want to pursue (except for extensions/hooks where it makes sense of course).

Comment (berrange) [inline pkg/virt-handler/vm.go]: I'm not debating whether the docker image feature makes sense - just the approach to implementation. There are countless different ways to do dynamic storage and this docker image is just one possibility. Adding code into the core to support every different approach does not make sense - these should be done as pluggable storage technologies via a generic framework to enable arbitrary third party implementations, of which this docker container approach is just one such impl.

Comment (davidvossel) [inline pkg/virt-handler/vm.go]: > Calling it "agile" is just using a buzzword to justify creating technical debt by not considering the general storage picture. 

Don't undermine my response by criticising the term agile. I could easily do the same for use of 'technical debt'. That doesn't further the discussion.

I didn't throw out a buzz word and leave it at that. I clearly outlined exactly what I meant by the term and stated the immediate tangible benefits.

> Adding code into the core to support every different approach does not make sense - these should be done as pluggable storage technologies via a generic framework to enable arbitrary third party implementations, of which this docker container approach is just one such impl.

In time. We have a real need for this feature now. In the future we will adapt as the generic storage abstraction progresses. We aren't there yet though.

Comment (rmohr) [inline pkg/virt-handler/vm.go]: > Adding code into the core to support every different approach does not make sense 

This seems to be a good candidate for the core. If we later have an abstraction we can move them behind it, if we really want.

I think that the concerns regarding a technical dept were addressed enough, and the arguments by @davidvossel and by me are explaining pretty well that there is no need to be concerned. So if there are no new arguments, I would appreciate if we could go back to discuss the details of the implementation.

**Comment by berrange** [comment]
No, i fundamentally disagree with making this a core component, and want to see the generic pluggable approach taken.

**Comment by rmohr** [comment]
So @berrange , to summarise again, what has already been said: This PR does not in any way block thoughts in having a common layout for doing storage. If your idea gains traction, get's accepted and it makes sense to move this kind of plugin behind this then developed abstraction, we will do it.

Right now, this PR implements a very well described use-case in #310, which no one is questioning. Also the user-facing API is pretty much independent from any implementation details. Also everyone agrees that we don't need a perfect solution from the start, and that we can improve it.

Could you please also try to answer to the brought up arguments in the comments before too? If you have reasonable concerns, then we definitely should consider them.

File: pkg/virt-handler/vm.go (hunk 1 above)
Comment (rmohr) [inline pkg/virt-handler/vm.go]: > There are countless different ways to do dynamic storage and this docker image is just one possibility

It is one, which is implicitly brought to us by using k8s. It is a little bit like PVs and PVCs. Whenever you deal with k8s, you either have a public or a private docker registry at your disposal.

**Comment by fabiand** [comment]
There are some disagreements around the approach and implementation.

I share @berrange s concerns. But OTOH we also see support for this implementation.

KubeVirt is still in an early phase: In general I see us still in the situation where we need to conduct research in many areas, trying out different approaches, to finally end up with an optimal solution considering the constraints and requirements.

I'd like to see that we use this freedom and experiment and eventually fail. But I'm also confident that we get back on track after identifying mistakes or finding a better path.

Hence I'd go forward and get this PR merged but also expect motivation for follow up fixes once they become necessary.

**Comment by berrange** [comment]
It is already clear that we need a generic mechanism for handling dynamic storage - my PR has already outlined multiple scenarios in which it is useful. If the proposal I've made is not satisfactory then provide an alternative suggestion that enables a generic mechanism for dynamic storage which doesn't involve code changes in virt-handler/virt-controller for each storage type. The proposal I've made is simple & straightforward to implement & can easily be done in this PR instead of taking a short term hack to rush in a feature that is not time critical to merge

**Comment by davidvossel** [comment]
sigh... I understand your point of view Dan... At this point I'm pretty sure you understand where I'm coming from as well. I disagree the generic storage api has to be coupled with this PR (for reasons I've already sited multiple times)... and so the circle goes. 

@berrange We're on the same team here. Please continue work on fleshing out and demonstrating the generic storage api so we can continue that discussion with you **outside** of this PR.  I think that's the way this is supposed to work. Someone makes a proposal that argues why a feature is necessary, they demonstrate the proposal with code, gain support/contributors, get merged... Holding another PR hostage to attempt to force someone else to demonstrate your proposal is not a solid plan for garnering support. I know that sounds harsh and isn't a perfect analogy, but I believe that is a fair representation of where this discussion is headed.  

The project isn't going to fall apart while we iterate on storage techniques. I believe in your ideas. Go make them a reality so we can make awesome things together.

**Comment by davidvossel** [comment]
updated the patches to work with new client-go changes and fixed merge conflicts in the hack/default-config file

File: tests/registry_disk_test.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	tests.PanicOnError(err)
+
+	restClient, err := kubecli.GetRESTClient()
+	tests.PanicOnError(err)
+
+	BeforeEach(func() {
+		tests.BeforeTestCleanup()
+	})
+
+	LaunchVM := func(vm *v1.VM) runtime.Object {
```
Comment (admiyo) [inline tests/registry_disk_test.go]: I think there is a reusable function for this already.

File: tests/registry_disk_test.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+		obj, err := restClient.Post().Resource("vms").Namespace(tests.NamespaceTestDefault).Body(vm).Do().Get()
+		Expect(err).To(BeNil())
+		return obj
+	}
+
+	VerifyVM := func(vm *v1.VM, obj runtime.Object) {
+		_, ok := obj.(*v1.VM)
```
Comment (admiyo) [inline tests/registry_disk_test.go]: this is a strange check.  you pass in the post-cast vm , but cast it again the confirm?

File: tests/registry_disk_test.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+		tests.BeforeTestCleanup()
+	})
+
+	LaunchVM := func(vm *v1.VM) runtime.Object {
+		obj, err := restClient.Post().Resource("vms").Namespace(tests.NamespaceTestDefault).Body(vm).Do().Get()
+		Expect(err).To(BeNil())
+		return obj
+	}
+
+	VerifyVM := func(vm *v1.VM, obj runtime.Object) {
```
Comment (admiyo) [inline tests/registry_disk_test.go]: can you give it a more explicit name and make reusable?  This is explicitly for verifying Registry disks, so maybe along those lines?

File: tests/registry_disk_test.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+				vms = append(vms, vm)
+				objs = append(objs, obj)
+			}
+
+			for idx, vm := range vms {
+				VerifyVM(vm, objs[idx])
+			}
+
+			close(done)
+		}, 45)
```
Comment (admiyo) [inline tests/registry_disk_test.go]: that is a long time out.  Can you make a symbolic constant and put the justification for it next to it?

File: pkg/registry-disk/registry-disk.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+//     INSERT DYNAMIC AUTH LOGIC HERE
+//}
+
+// The controller applies ports to registry disks when a VM spec is introduced into the cluster.
+func ApplyPorts(vm *v1.VM) {
+	wrapperStartingPort := 3261
+	for idx, disk := range vm.Spec.Domain.Devices.Disks {
+		if disk.Type == registryDiskV1Alpha {
+			port := fmt.Sprintf("%d", wrapperStartingPort)
+			name := "127.0.0.1"
```
Comment (admiyo) [inline pkg/registry-disk/registry-disk.go]: why is this hard coded?

File: tests/registry_disk_test.go (hunk 2 above)
Comment (davidvossel) [inline tests/registry_disk_test.go]: I didn't find one. Most of the other tests are doing a synchronized "launch this vm and wait for it". This function is different. We want to launch multiple VMs in parallel.  That logic doesn't exist anywhere else in the test suite yet.

If we end up reusing this type of logic outside of registry disks, we can make it generic.

File: tests/registry_disk_test.go (hunk 4 above)
Comment (davidvossel) [inline tests/registry_disk_test.go]: I renamed this in an update that's i'm about to push. It's only reusable within the context of registry disks.

File: tests/registry_disk_test.go (hunk 3 above)
Comment (davidvossel) [inline tests/registry_disk_test.go]: the runtime obj is used for the shared "WaitForSuccessfulVMStart" function. I do the cast just to confirm the obj type before going any further.

File: tests/registry_disk_test.go (hunk 5 above)
Comment (davidvossel) [inline tests/registry_disk_test.go]: i could make it a symbolic constant, but there's no precedence for that. We have other tests that require longer timeouts then this. 

I put a comment explaining why the long timeout in the update that I'm pushing soon.

File: pkg/registry-disk/registry-disk.go (hunk 6 above)
Comment (davidvossel) [inline pkg/registry-disk/registry-disk.go]: 127.0.0.1 is the default host until it is overwritten by the ip associated with the scheduled pod. In my update I'm making this more clear.

**Comment by davidvossel** [comment]
alrighty, I rebased the patches, addressed @admiyo comments, and included some new enhancements related to readiness checks.

I also added a registry disk for fedora atomic to the repo. Since the qcow2 image is around 700mb, I made this image optional.

To run fedora atomic with an ephemeral registry disk run the following
```
make vagrant-sync-optional
cluster/kubectl create -f cluster/vm-atomic.yaml
```

The atomic image is something I'm using to prove another patch series related to cloud-init that I'll be submitting a PR for soon.

If we're over the discussion phase on this feature, I'd encourage the team to merge this. It's about to start blocking other PRs.

File: cmd/registry-disk-v1alpha/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+#
+
+FROM debian:sid
+
+MAINTAINER "David Vossel" \<dvossel@redhat.com\>
+ENV container docker
+
+RUN apt-get update && apt-get install -y \
+	tgt bash curl bzip2 qemu-utils && \
+	mkdir -p /disk
```
Comment (rmohr) [inline cmd/registry-disk-v1alpha/Dockerfile]: Adding `rm -rf /var/lib/apt/lists/*` should also save us a little bit of disk space

File: pkg/virt-controller/watch/vm.go [hunk 8]
```diff
@@ -226,8 +230,17 @@ func (c *VMController) execute(key string) error {
 			return nil
 		}
 
+		// Ensure registry disks are online before placing VM
+		if registrydisk.DisksAreReady(&pods.Items[0]) == false {
+			return errors.New("waiting on image wrapper disks to become ready")
```
Comment (rmohr) [inline pkg/virt-controller/watch/vm.go]: In that case the appropriate thing would be to log that we are waiting for the image to be ready. It can take some time. It is not an error, therefore we need no automatic re-queue. After all disks are ready, we get notified by the apiserver because of the pod state change. I would log it as info with a severity > 1 (we should somwhere define what log level to use for what ...) and return nil.

File: tests/registry_disk_test.go [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+	restClient, err := kubecli.GetRESTClient()
+	tests.PanicOnError(err)
+
+	BeforeEach(func() {
+		tests.BeforeTestCleanup()
+	})
+
+	LaunchVM := func(vm *v1.VM) runtime.Object {
+		obj, err := restClient.Post().Resource("vms").Namespace(tests.NamespaceTestDefault).Body(vm).Do().Get()
```
Comment (rmohr) [inline tests/registry_disk_test.go]: We can already use the kubevirt client, instead of the plain rest client. Something like (cli.VM(tests.NamespaceTestDefault).Create(VM)) should be the call.

File: tests/registry_disk_test.go [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+		Expect(err).To(BeNil())
+		return obj
+	}
+
+	VerifyRegistryDiskVM := func(vm *v1.VM, obj runtime.Object) {
+		_, ok := obj.(*v1.VM)
```
Comment (rmohr) [inline tests/registry_disk_test.go]: It would also be acceptible for me to do an unconditional cast, and just let  the test panic.

File: tests/registry_disk_test.go [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+				vms = append(vms, vm)
+				objs = append(objs, obj)
+			}
+
+			for idx, vm := range vms {
+				VerifyRegistryDiskVM(vm, objs[idx])
+			}
+
+			close(done)
+		}, 45) // Timeout is long because this test involves multiple parallel VM launches.
```
Comment (rmohr) [inline tests/registry_disk_test.go]: The timout is fine, but independent of that, we should really increase our worker count in the controller and virt-handler. It is currently 1, but we can cange it to 3 or 5. One of these numbers is also what k8s uses. Let me check that.

File: pkg/registry-disk/registry-disk.go [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+	}
+	return true
+}
+
+// The virt-handler converts registry disks to their corresponding iscsi network
+// disks when the VM spec is being defined as a domain with libvirt.
+// The ports and host of the iscsi disks are already provided here by the controller.
+func MapRegistryDisks(vm *v1.VM) (*v1.VM, error) {
+	vmCopy := &v1.VM{}
+	model.Copy(vmCopy, vm)
```
Comment (rmohr) [inline pkg/registry-disk/registry-disk.go]: Hm, at some point, it is better, to just copy the vm object once, when we read it the first time out of the cache in the loop.

File: tests/registry_disk_test.go (hunk 9 above)
Comment (rmohr) [inline tests/registry_disk_test.go]: In general, when using the rest client, there is a nicer pattern, instead of doing explicit calls, I forget that myself pretty often:

```
vm := &v1.VM{}
Expect(restClient.Post().Resource("vms").Namespace(tests.NamespaceTestDefault).Body(vm).Do().Into(vm)).To(Succeed())
return vm

```

File: cluster/vm-ephemeral.json [hunk 13]
```diff
@@ -0,0 +1,42 @@
+{
```
Comment (davidvossel) [inline cluster/vm-ephemeral.json]: removed, in the update i'm pushing soon

File: cluster/vm-atomic.yaml [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+          network: default
+        type: network
+    memory:
+      unit: MB
+      value: 1024
+    os:
+      type:
+        os: hvm
+    type: qemu
+  cloudInit:
```
Comment (davidvossel) [inline cluster/vm-atomic.yaml]: yep yep, my cloud-init functional tests actually use cirros.   I need to check to see if cirros actually uses the cloud-init method i'm using though. It's possible i'm just attaching a file disk to cirros with  userdata that cirros ignores.

File: cmd/registry-disk-v1alpha/Dockerfile (hunk 7 above)
Comment (davidvossel) [inline cmd/registry-disk-v1alpha/Dockerfile]: fixed

File: pkg/virt-controller/watch/vm.go (hunk 8 above)
Comment (davidvossel) [inline pkg/virt-controller/watch/vm.go]: good catch. I fixed this.

File: tests/registry_disk_test.go (hunk 9 above)
Comment (davidvossel) [inline tests/registry_disk_test.go]: I agree there's room for improvement here. Part of the reason I didn't start using the kubevirt client is because helper functions expect the runtime obj to be passed in. 

I've noted to follow up on this in a separate patch. To do it right, this is going to involve a more broad set of changes to our functional tests. The code is the way it is in the registry disk tests because it is consistent with the other functional tests and the common utils.go file.

File: pkg/registry-disk/registry-disk.go (hunk 12 above)
Comment (davidvossel) [inline pkg/registry-disk/registry-disk.go]: yep agreed. I noted that for another patch. I don't think I'm going to tackle that in this pr though.

**Comment by davidvossel** [comment]
patch updated to address @rmohr feedback. Includes controller unit test to exercise registry disk logic.

**Comment by davidvossel** [comment]
> We also need to add the possibility to specify the pullPolicy per disk. Kubernetes support Always, IfNotPresent and Never. However, that is fine for me, if it is introduced later.

Good idea. I've noted that on the list of things I want to follow up on after this patch is merged.
